builtin/game/item_entity.lua

```diff
@@ -318,15 +318,50 @@ core.register_entity(":__builtin:item", {
 		end
 	end,
 
-	on_punch = function(self, hitter)
-		local inv = hitter:get_inventory()
-		if inv and self.itemstring ~= "" then
-			local left = inv:add_item("main", self.itemstring)
-			if left and not left:is_empty() then
-				self:set_item(left)
+	on_punch = function(self, hitter, ...)
+		if self.itemstring ~= "" then
+			-- Call on_pickup callback in item definition.
+			local itemstack = ItemStack(self.itemstring)
+			local itemstack_modified = false
+			local item_def = itemstack:get_definition()
+			if item_def and item_def.on_pickup then
+				local ret = item_def.on_pickup(ItemStack(itemstack), hitter,
+						self.object, ...)
+				if not ret then
+					return
+				elseif ret ~= true then
+					itemstack_modified = true
+					itemstack = ItemStack(ret)
+				end
+			end
+
+			-- Invoke global on_item_pickup callbacks.
+			if not core.run_callbacks(core.registered_on_item_pickups, 3,
+					ItemStack(itemstack), hitter, self.object, ...) then
+				return
+			end
```

Comment:
What would you think of making this similar to `minetest.item_eat`? If added to the itemdef metatable, the function could serve as a default for all picked up items.

Just thinking out loud because these two look very similar, and having less different concepts would be great to have. Like this, you could use `on_pickup = minetest.item_eat(2),` in an ItemDef to eat by picking up.

Reply:
I've changed the obj parameter to a `pointed_thing`, so `on_pickup = minetest.item_eat(2),` should work now.

In that sense, it is now more similar to `on_use`.
Idk. how I should make it more similar to `item_eat`.